File: apps/mobile/lib/features/home/domain/entities/feed_item.dart
/// Feed item entity for home feed display
///
/// PROJECTION LAYER ONLY: This is a UI projection entity, NOT domain truth.
/// Source of truth is the canonical Content entity from Content domain.
///
/// R3.1: Created to replace missing shared FeedItem entity
/// This entity represents a single item in the user's home feed
///
/// CONTRACT:
/// - Only universal social content appears in home feed
/// - Commerce types (auction) are filtered by backend
/// - Engagement counts are provided (may be 0 if not supported by backend)
///
/// MEDIA INTEGRATION: Uses canonical MediaEntity from Content domain
/// Feed projection may have simplified media - resolve full Content for complete media data
library;

import 'package:labuda/domains/social/content/domain/entities/content.dart'; // For MediaEntity
import 'package:labuda/shared/governance/content_lifecycle.dart';

/// Feed item types
enum FeedItemType {
  /// Universal social content item
  content,

  /// P3A — Promoted listing injected by backend feed promotion injector.
  promotedListing,

  /// P3A — Promoted auction injected by backend feed promotion injector.
  promotedAuction,

  /// P3A — Promoted external product injected by backend feed promotion injector.
  promotedExternal,
}

/// R3.1: Extension to add displayName to FeedItemType
extension FeedItemTypeExtension on FeedItemType {
  /// Get display name for UI
  String get displayName {
    switch (this) {
      case FeedItemType.content:
        return 'Content';
      case FeedItemType.promotedListing:
        return 'Dipromosikan';
      case FeedItemType.promotedAuction:
        return 'Dipromosikan';
      case FeedItemType.promotedExternal:
        return 'Dipromosikan';
    }
  }
}

/// Feed item entity
///
/// PROJECTION LAYER: UI view of Content entity for feed display.
/// This is NOT the source of truth - always resolve canonical Content for business logic.
///
/// Represents a single content item in the user's home feed.
/// Used across home and profile features for consistent feed display.
///
/// MEDIA INTEGRATION: Uses `List<MediaEntity>` instead of `List<String>` mediaUrls
/// Source of truth: canonical Content.media from content_media table
class FeedItem {
  final String id;
  final String content;
  final String authorId;
  final String? authorUsername;
  final String? authorAvatarUrl;
  final FeedItemType type;
  final DateTime createdAt;

  /// MEDIA INTEGRATION: Media from canonical Content.media
  /// Contract: Sourced from content_media table via backend Feed API
  final List<MediaEntity> media;

  final int likes;
  final int comments;
  final List<String> likedByUsers;
  final Map<String, dynamic> additionalData;

  /// Canonical governance lifecycle. Defaults to active for null/unknown
  /// wire values so legacy payloads keep rendering today's behavior.
  final ContentLifecycle lifecycle;

  /// E2.1 — Canonical governance lifecycle for the AUTHOR identity
  /// (PublicCard.UserCard.Lifecycle). Independent of [lifecycle]; the
  /// author may be unavailable/removed while the content itself is still
  /// active. Defaults to active for null / unknown / missing wire values
  /// so legacy payloads keep rendering today's behavior.
  ///
  /// Active feed surface only — non-feed surfaces (chat, comments,
  /// notifications, profile, search) still receive null and degrade to
  /// active per the same fallback.
  final ContentLifecycle authorLifecycle;

  /// FIX-3 — Canonical governance lifecycle for the ORIGINAL AUTHOR of a
  /// repost. Sourced from `original_author_lifecycle` on the feed wire.
  /// Only meaningful when [isRepost] is true. Defaults to active for null /
  /// unknown / missing wire values (legacy reposts degrade safely to current
  /// behavior). When degraded, attribution display shows a placeholder.
  final ContentLifecycle originalAuthorLifecycle;

  const FeedItem({
    required this.id,
    required this.content,
    required this.authorId,
    this.authorUsername,
    this.authorAvatarUrl,
    required this.type,
    required this.createdAt,
    this.media = const [],
    this.likes = 0,
    this.comments = 0,
    this.likedByUsers = const [],
    this.additionalData = const {},
    this.lifecycle = ContentLifecycle.active,
    this.authorLifecycle = ContentLifecycle.active,
    this.originalAuthorLifecycle = ContentLifecycle.active,
  });

  /// Get title from additional data
  String? get title => additionalData['title'] as String?;

  /// Get caption from additional data
  String? get caption => additionalData['caption'] as String?;

  /// Check if item is hidden
  bool get isHidden => additionalData['isHidden'] as bool? ?? false;

  /// Get status from additional data
  String? get status => additionalData['status'] as String?;

  /// Check if this is a repost
  bool get isRepost => additionalData['isRepost'] as bool? ?? false;

  /// Get original author ID for reposts
  String? get originalAuthorId => additionalData['originalAuthorId'] as String?;

  /// Check if item has media
  bool get hasMedia => media.isNotEmpty;

  /// Create a copy with modified fields
  FeedItem copyWith({
    String? id,
    String? content,
    String? authorId,
    String? authorUsername,
    String? authorAvatarUrl,
    FeedItemType? type,
    DateTime? createdAt,
    List<MediaEntity>? media,
    int? likes,
    int? comments,
    List<String>? likedByUsers,
    Map<String, dynamic>? additionalData,
    ContentLifecycle? lifecycle,
    ContentLifecycle? authorLifecycle,
    ContentLifecycle? originalAuthorLifecycle,
  }) {
    return FeedItem(
      id: id ?? this.id,
      content: content ?? this.content,
      authorId: authorId ?? this.authorId,
      authorUsername: authorUsername ?? this.authorUsername,
      authorAvatarUrl: authorAvatarUrl ?? this.authorAvatarUrl,
      type: type ?? this.type,
      createdAt: createdAt ?? this.createdAt,
      media: media ?? this.media,
      likes: likes ?? this.likes,
      comments: comments ?? this.comments,
      likedByUsers: likedByUsers ?? this.likedByUsers,
      additionalData: additionalData ?? this.additionalData,
      lifecycle: lifecycle ?? this.lifecycle,
      authorLifecycle: authorLifecycle ?? this.authorLifecycle,
      originalAuthorLifecycle:
          originalAuthorLifecycle ?? this.originalAuthorLifecycle,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is FeedItem && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() {
    return 'FeedItem(id: $id, type: $type, authorId: $authorId, createdAt: $createdAt, mediaCount: ${media.length})';
  }
}

File: backend/internal/social/feed/entity/feed_item.go
// DOMAIN: SOCIAL
// NOTE: Aggregated content feed for timeline display

package entity

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// FeedMedia represents a media attachment for feed items.
// Source of truth: canonical Content.media (List<MediaEntity>)
//
// PHASE C — MediaRef convergence (Option D, transitional hybrid).
// The legacy fields URL/Type/Position are preserved for mobile
// parser compatibility (FeedMediaDto in apps/mobile/lib/features/
// home/data/dto/feed_dto.dart binds them as required non-null).
// The new Kind/Width/Height fields make this struct serializable
// as a canonical-compatible MediaRef without renaming or removing
// legacy fields. See backend/internal/pkg/mediaref/mediaref.go for
// the canonical contract.
//
// Hydration rules in this layer:
//   - Kind mirrors Type (set in feed_repository_impl after JSON
//     unmarshal of the per-row aggregate). Never nil in practice
//     once hydrated; nil only if the entity is constructed without
//     going through repository hydration.
//   - Width / Height are always nil in this layer — no DB column,
//     no inference, no remote fetch.
type FeedMedia struct {
	URL      string `json:"url"`
	Type     string `json:"type"` // "image" or "video"
	Position int    `json:"position"`

	// Canonical-compatible additive fields (Phase C). Always
	// emitted (no omitempty) so the wire shape is canonical-
	// compatible; null until hydrated.
	Kind   *string `json:"kind"`
	Width  *int    `json:"width"`
	Height *int    `json:"height"`
}

// FeedItem represents a single content item in the feed.
//
// SCHEMA ALIGNMENT (Batch 3J): the canonical contents table has no
// `body` / `title` columns — only `caption`. We keep the `Body`
// field on this projection entity because the public API contract
// still carries `body` (mobile requires it); the field is populated
// from `c.caption` via SQL alias in the repository layer. The
// legacy `Title` field has been removed — it was scanned from a
// nonexistent column and never made it into any live response.
type FeedItem struct {
	ID       uuid.UUID
	AuthorID uuid.UUID
	Type     string
	// Status is always "active" for feed items
	Status    string
	Body      string  // sourced from contents.caption via SQL alias
	Caption   *string // canonical text column on contents
	City      *string // POST LOCATION: City name for posts only
	Province  *string // POST LOCATION: Province name for posts only
	IsHidden  bool
	CreatedAt time.Time
	UpdatedAt time.Time

	// Author information (optional, for display)
	AuthorUsername *string
	AuthorAvatar   *string
	AuthorCity     *string // Author's city from user_profiles
	AuthorProvince *string // Author's province from user_profiles

	// AuthorLifecycle is the coarsened public user lifecycle for the
	// content author, sourced from users.account_status + users.deleted_at
	// at projection time and coarsened via
	// viewercontext.CoarsenLifecycle (Go-side; never via SQL).
	//
	// Empty string means lifecycle was not hydrated (legacy fallback /
	// rollback safety). Non-empty values are constrained to the canonical
	// public lifecycle vocabulary: "active", "unavailable", "removed".
	//
	// Per E2 doctrine (feed identity transport activation), this field is
	// the SOLE feed-surface carrier for the embedded UserCard.Lifecycle
	// public emission. No other surface is wired through this seam yet.
	AuthorLifecycle string

	// SHARE CONTRACT V1: Repost attribution fields
	// If non-nil, this item is a repost/shares of another content
	OriginalAuthorID *uuid.UUID

	// MEDIA INTEGRATION: Media from canonical Content.media
	// Contract: Must be sourced from content_media table
	// NO hardcoded empty arrays, NO dummy data
	Media []FeedMedia `json:"media"`
}

// FeedResult contains the feed items and pagination cursor.
//
// NextCursor is the opaque, base64-encoded continuation cursor of the
// last returned row; nil when HasMore is false. HasMore is derived
// from a LIMIT+1 probe in the repository, not from len(Items).
type FeedResult struct {
	Items      []*FeedItem
	NextCursor *FeedCursor
	HasMore    bool
}

// FeedCursor is the opaque pagination cursor for the feed surface.
//
// It identifies a row in the canonical (created_at DESC, id DESC)
// ordering by the (created_at, id) tuple of the boundary row that has
// already been returned to the client. The cursor is treated as opaque
// at the HTTP boundary — clients must not parse or interpret it, and
// the wire format may change without notice.
//
// Use [EncodeFeedCursor] / [DecodeFeedCursor] for the canonical string
// form crossed at the HTTP boundary.
type FeedCursor struct {
	PriorityGroup int       `json:"pg"`
	CreatedAt     time.Time `json:"ts"`
	ID            uuid.UUID `json:"id"`
}

// EncodeFeedCursor returns the canonical opaque string form of c.
//
// The encoding is URL-safe base64 (no padding) of the JSON payload
// `{"ts":"<RFC3339Nano>","id":"<uuid>"}`. Returns "" for nil so the
// caller can decide between omitting the field or emitting an empty
// string; the handler emits the encoded value only when HasMore=true.
func EncodeFeedCursor(c *FeedCursor) string {
	if c == nil {
		return ""
	}
	payload, err := json.Marshal(c)
	if err != nil {
		// json.Marshal cannot fail for a struct with these field types,
		// but defend against future shape drift by returning empty —
		// the handler treats "" as "no cursor available".
		return ""
	}
	return base64.RawURLEncoding.EncodeToString(payload)
}

// DecodeFeedCursor parses an opaque cursor string back into a
// FeedCursor.
//
// An empty input returns (nil, nil) — the canonical "no cursor" form
// used for the first page. A malformed cursor returns a non-nil error
// so the caller (HTTP handler) can map it to 400.
//
// BACKWARD COMPATIBILITY: the prior cursor format was an RFC3339Nano
// timestamp string. There is no production data on this surface yet
// (clean-break doctrine, MEMORY: feedback_from_zero_no_backcompat), so
// the legacy format is intentionally rejected. Mobile builds carrying
// an old cursor in memory will receive 400 once on the first page-2
// request after upgrade and recover via the standard refresh path.
func DecodeFeedCursor(s string) (*FeedCursor, error) {
	if s == "" {
		return nil, nil
	}
	raw, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		return nil, fmt.Errorf("decode feed cursor: %w", err)
	}
	var c FeedCursor
	if err := json.Unmarshal(raw, &c); err != nil {
		return nil, fmt.Errorf("decode feed cursor: %w", err)
	}
	if c.CreatedAt.IsZero() {
		return nil, fmt.Errorf("decode feed cursor: missing ts")
	}
	if c.ID == uuid.Nil {
		return nil, fmt.Errorf("decode feed cursor: missing id")
	}
	return &c, nil
}
